src/centrality.rs

```diff
@@ -0,0 +1,217 @@
+use pyo3::prelude::*;
+
+use petgraph::graph::NodeIndex;
+use petgraph::visit::{
+    GraphBase,
+    GraphProp, // allows is_directed
+    IntoNeighborsDirected,
+    IntoNodeIdentifiers,
+    NodeCount,
+    NodeIndexable,
+};
+use std::collections::HashMap;
+
+// Correspondence to notation in Brandes 2001
+//
+// P -- predecessors
+// S -- sorted_by_distance,
+//      vertices in order of non-decreasing distance from s
+// Q -- Q
+// sigma -- sigma
+// delta -- delta
+// d -- distance
+
+pub fn betweenness_centrality<G>(
+    graph: G,
+    endpoints: bool,
+    normalized: bool,
+) -> PyResult<Vec<f64>>
+where
+    G: NodeIndexable
+        + IntoNodeIdentifiers
+        + IntoNeighborsDirected
+        + NodeCount
+        + GraphProp
+        + GraphBase<NodeId = NodeIndex>,
+    // rustfmt deletes the following comments if placed inline above
+    // + IntoNodeIdentifiers // for node_identifiers()
+    // + IntoNeighborsDirected // for neighbors()
+    // + NodeCount // for node_count
+    // + GraphProp // for is_directed
+{
+    let mut betweenness = vec![0.0; graph.node_count()];
```

Comment:
So I assume based on this betweeness is being returned as a `Vec` where the index in the Vec matches the node index in the graph. The problem with this is that there is no guarantee with the stable graph struct we use for the node indices to be a contiguous set, we can end up with holes in the indices list if nodes are removed. So there are situations where `max(node_indices) != max(node_count)`.

So I'm thinking it might make sense to change the output type to be a `HashMap<usize, f64>` instead. You can use an iterator and collect to intialize it in the same way. Something like:

```rust
let mut betweeness: HashMap<usize, f64> = graph.node_identifiers().map(|iden| (graph.to_index(iden), 0.0)).collect();
```
```

Reply:
One thing here I should have mentioned is that if you don't want to use a hashmap for performance you could use `graph.node_bound()` here to get the max node index in the graph and use that instead. But if you were I would probably switch this to be `Vec<Option<f64>>` instead so we can set any potential holes in the indices to `None`. We probably should still return a `HashMap` for ease of accessing from python, but that's as easy as something like: `out_map: HashMap<usize, f64> = betweeness.iter().enumerate().filter(|(index, v)| v.is_some()).collect()` to filter out None and return a HashMap.

Reply:
> you could use `graph.node_bound()` here to get the max node

I don't understand how this solves the problem of holes in the node index set.

> I would probably switch this to be `Vec<Option<f64>>`

This seems to amount to looping over all integers, checking whether each one is in the index set. Is it reasonable that someone would have, say,  five nodes only with contiguous indices starting at `10^6`? In this case, using `Vec<Option<f64>>` far less space and time efficient than using a hashmap.

I am in the process of implementing the `Vec<Option<f64>>` idea. In fact it's done, but buggy; betweenness values are wrong. But, as I am finishing, it occurs to me that this may not be  a good solution.

In any case, I agree that returning a `dict` is a good idea.

Reply:
I have changed the implementation to use `Vec<Option<f64>>` and pushed the commits.